scripts/mkstatic.py

```diff
@@ -915,29 +926,109 @@ def write(self, metadata_request, metadata_define, arguments):
                     error_check = ''
                     args = ''
                     length = 0
-                    # Extract all variables needed (including indices for components/slices of arrays)
+
+                    # First identify all dimensions needed to handle the arguments
+                    # and add them to the list of required variables for the cap
+                    additional_variables_required = []
                     for var_standard_name in arguments[module_name][scheme_name][subroutine_name]:
+                        if not var_standard_name in metadata_define.keys():
+                            raise Exception('Variable {standard_name} not defined in host model metadata'.format(
+                                                                                standard_name=var_standard_name))
+                        var = metadata_define[var_standard_name][0]
+                        # dim can be 'A', '1', '1:A', ...
+                        for dim_expression in var.dimensions:
+                            dims = dim_expression.split(':')
+                            for dim in dims:
+                                try:
+                                    dim = int(dim)
+                                except ValueError:
+                                    if not dim in local_vars.keys() and \
+                                            not dim in additional_variables_required + arguments[module_name][scheme_name][subroutine_name]:
+                                        logging.debug("Adding dimension {} for variable {}".format(dim, var_standard_name))
+                                        additional_variables_required.append(dim)
+
+                        # If blocked data structures need to be converted, add necessary variables
+                        if ccpp_stage in ['init', 'finalize'] and CCPP_INTERNAL_VARIABLES[CCPP_BLOCK_NUMBER] in var.local_name:
+                            if not CCPP_BLOCK_COUNT in local_vars.keys() \
+                                    and not CCPP_BLOCK_COUNT in additional_variables_required + arguments[module_name][scheme_name][subroutine_name]:
+                                    logging.debug("Adding variable {} for handling blocked data structures".format(CCPP_BLOCK_COUNT))
+                                    additional_variables_required.append(CCPP_BLOCK_COUNT)
+                            if not CCPP_HORIZONTAL_LOOP_EXTENT in local_vars.keys() \
+                                    and not CCPP_HORIZONTAL_LOOP_EXTENT in additional_variables_required + arguments[module_name][scheme_name][subroutine_name]:
+                                    logging.debug("Adding variable {} for handling blocked data structures".format(CCPP_HORIZONTAL_LOOP_EXTENT))
+                                    additional_variables_required.append(CCPP_HORIZONTAL_LOOP_EXTENT)
```

Comment:
This is a tricky one. Because the copying of the blocked data structure requires knowledge about the block sizes, for example (this is from the auto-generated code):
```
        allocate(tmpvar17(1:GFS_Control%ncols))
        ib = 1
        do nb=1,GFS_Control%nblks
          tmpvar17(ib:ib+GFS_Control%blksz(nb)-1) = GFS_Data(nb)%Grid%area
          ib = ib+GFS_Control%blksz(nb)
        end do
```
`GFS_Control%blksz(nb)` is the variable corresponding to `horizontal_loop_extent`.

Reply:
Oh duh, thanks for the explanation.